### DISCUSSION TITLE
Snapping to "Center Half" doesn't continue to "Center Third" with subsequent presses

### DISCUSSION THREAD

**Discussion by ZeeEssDoubleU** [rxhanson/Rectangle#466] [General]
App does not snap to "Center Third" after snapping to "Center Half" on subsequent key presses.  I think this should behave like the other "Half" options (ie "Left Half"), where they snap to their "Third" and "Two Thirds" variants if another hotkey isn't set for those respective options.

macOS version: 11.3 (20E232)
Rectangle version: 0.45 (50)

**Comment by rxhanson** [comment]
Currently this is enabled via a hidden config:

defaults write com.knollsoft.Rectangle centerHalfCycles -int 1

Followed by an app restart.

I'll probably enable it by default in a future release - it's just that it got added in much later and I wasn't sure if I wanted it to be the default behavior initially.

**Reply by densh** [reply]
I'd love to see this become the default.

**Comment by aviflax** [comment]
I’m having trouble getting this to work. I ran the command in Terminal and restarted the app.

MacOS 11.4 (20F71)
Rectangle 0.46 (51)

Any suggestions?

Thanks!

**Reply by rxhanson** [reply]
I'll go ahead and just make it the default in the next release, which should be out sometime this week.

**Reply by aviflax** [reply]
Great, thanks!

**Reply by aviflax** [reply]
I just updated to 0.47 and was initially puzzled when the feature _still_ didn’t seem to work for me. So I poked around in the preferences again, and this time I noticed that under _Move to adjacent display on repeated left or right commands_ there’s some light gray text that reads _When unchecked, repeated half actions cycle ½, ⅔, and ⅓_ — so I unchecked, and now the feature works for me.

I’m not sure if there’s a way to make the interaction of these features+settings clearer, it does seem quite complicated!

Anyway, I'm enjoying this feature very much. Thank you!

**Reply by rxhanson** [reply]
Good point, thanks for letting me know. In hindsight, that's probably why the terminal command didn't work for you either. 

I agree, the messaging should be more clear here. I have some thoughts on how to improve this, it's just a matter of determining when I want to make changes.

**Reply by jfmcoronel** [reply]
FWIW, I ran into the same issue today. I was trying to figure out why the `centerHalfCycles` setting was not working after seeing in the readme that it was available.

Maybe we could have a sentence in the readme saying that the _When unchecked, repeated half actions cycle ½, ⅔, and ⅓_ checkbox has to be unticked for `centerHalfCycles` to take effect.

**Reply by rxhanson** [reply]
I just checked in a change that will honor the `centerHalfCycles` being enabled even if you don't have that checkbox unchecked. 

As a side note, the only weirdness that I left in is the scenario that you have manually disabled `centerHalfCycles` but have that box unchecked, it won't honor the disabling of 'centerHalfCycles' mainly because I don't think that's a scenario anyone will want (if you want the cycling at all, then I don't foresee anyone not wanting it for center half). Keeps the code cleaner.

This change will be out in the next release, hopefully soon.

**Reply by jfmcoronel** [reply]
Nice, that's even better for me.

I was used to repeatedly pressing `Left Half` and `Right Half` in order to move windows across multiple monitors, and having the _When unchecked, repeated half actions cycle ½, ⅔, and ⅓_ checkbox ticked made it cycle through sizes instead. I had to settle binding `Next Display` and `Previous Display` hotkeys as a workaround. Glad to know I could get the best of both worlds in the next release.

Thanks a lot for this and for the amazing tool!

**Comment by danielhoherd** [comment]
Is `defaults write com.knollsoft.Rectangle centerHalfCycles -int 1` still working in v0.90? I cannot seem to get it to work, and I would love to have a feature that cycles through different sizes like pretty much every other snap location does.

**Reply by rxhanson** [reply]
It's possible that the Terminal command was broken a while back when the ability to select from more cycling fractions was added. You should get center half cycling if you make sure the Repeated commands setting is set to cycle sizes on half actions:

<img width="602" height="786" alt="image" src="https://github.com/user-attachments/assets/aa654e98-d501-42f6-ae4f-9ba47e906ea0" />

**Reply by danielhoherd** [reply]
@rxhanson I set up my prefs exactly like yours are in that screenshot, including going to system prefs and setting "Window title bar double-click action: No Action", and I still do not see anything happen after the first center snap when I press ctrl-opt-c (my center snap shortcut) many times.

In case it helps, here is my RectangleConfig.json run through `jq -c`

```json
{"bundleId":"com.knollsoft.Rectangle","defaults":{"SUEnableAutomaticChecks":{"bool":true},"allowAnyShortcut":{"bool":true},"almostMaximizeHeight":{"float":0},"almostMaximizeWidth":{"float":0},"altThirdCycle":{"int":0},"alternateDefaultShortcuts":{"bool":false},"alwaysAccountForStage":{"int":0},"applyGapsToMaximize":{"int":0},"applyGapsToMaximizeHeight":{"int":0},"attemptMatchOnNextPrevDisplay":{"int":0},"autoMaximize":{"int":0},"cascadeAllDeltaSize":{"float":30},"centerHalfCycles":{"int":1},"centeredDirectionalMove":{"int":0},"cornerSnapAreaSize":{"float":20},"curtainChangeSize":{"int":0},"cycleSizesIsChanged":{"bool":true},"disabledApps":{},"doubleClickTitleBar":{"int":3},"doubleClickTitleBarIgnoredApps":{},"doubleClickTitleBarRestore":{"int":0},"dragFromStage":{"int":0},"enhancedUI":{"int":1},"footprintAlpha":{"float":0.3},"footprintAnimationDurationMultiplier":{"float":0},"footprintBorderWidth":{"float":2},"footprintColor":{},"footprintFade":{"int":0},"fullIgnoreBundleIds":{},"gapSize":{"float":0},"hapticFeedbackOnSnap":{"int":0},"hideMenubarIcon":{"bool":false},"ignoreDragSnapToo":{"int":0},"ignoredSnapAreas":{"int":0},"landscapeSnapAreas":{},"launchOnLogin":{"bool":true},"minimumWindowHeight":{"float":0},"minimumWindowWidth":{"float":0},"missionControlDragging":{"int":0},"missionControlDraggingAllowedOffscreenDistance":{"float":25},"missionControlDraggingDisallowedDuration":{"int":250},"moveCursor":{"int":0},"moveCursorAcrossDisplays":{"int":0},"notifiedOfProblemApps":{"bool":false},"obtainWindowOnClick":{"int":0},"portraitSnapAreas":{},"relaunchOpensMenu":{"bool":false},"resizeOnDirectionalMove":{"bool":false},"screenEdgeGapBottom":{"float":0},"screenEdgeGapLeft":{"float":0},"screenEdgeGapRight":{"float":0},"screenEdgeGapTop":{"float":0},"screenEdgeGapTopNotch":{"float":0},"screenEdgeGapsOnMainScreenOnly":{"bool":false},"screensOrderedByX":{"int":0},"selectedCycleSizes":{"int":7},"shortEdgeSnapAreaSize":{"float":145},"showAllActionsInMenu":{"int":0},"sixthsSnapArea":{"int":0},"sizeOffset":{"float":0},"snapEdgeMarginBottom":{"float":5},"snapEdgeMarginLeft":{"float":5},"snapEdgeMarginRight":{"float":5},"snapEdgeMarginTop":{"float":5},"snapModifiers":{"int":0},"specifiedHeight":{"float":1050},"specifiedWidth":{"float":1680},"stageSize":{"float":190},"subsequentExecutionMode":{"int":0},"systemWideMouseDown":{"int":0},"systemWideMouseDownApps":{},"todo":{"int":1},"todoApplication":{},"todoMode":{"bool":false},"todoSidebarSide":{"int":2},"todoSidebarWidth":{"float":7},"traverseSingleScreen":{"int":0},"unsnapRestore":{"int":2},"windowSnapping":{"int":2}},"shortcuts":{"almostMaximize":{"keyCode":36,"modifierFlags":917504},"bottomHalf":{"keyCode":125,"modifierFlags":786432},"bottomLeft":{"keyCode":38,"modifierFlags":786432},"bottomRight":{"keyCode":40,"modifierFlags":786432},"center":{"keyCode":8,"modifierFlags":786432},"centerThird":{"keyCode":3,"modifierFlags":786432},"firstThird":{"keyCode":2,"modifierFlags":786432},"firstTwoThirds":{"keyCode":14,"modifierFlags":786432},"larger":{"keyCode":124,"modifierFlags":917504},"lastThird":{"keyCode":5,"modifierFlags":786432},"lastTwoThirds":{"keyCode":17,"modifierFlags":786432},"leftHalf":{"keyCode":123,"modifierFlags":786432},"maximize":{"keyCode":36,"modifierFlags":786432},"maximizeHeight":{"keyCode":126,"modifierFlags":917504},"nextDisplay":{"keyCode":124,"modifierFlags":1835008},"previousDisplay":{"keyCode":123,"modifierFlags":1835008},"rightHalf":{"keyCode":124,"modifierFlags":786432},"smaller":{"keyCode":123,"modifierFlags":917504},"topHalf":{"keyCode":126,"modifierFlags":786432},"topLeft":{"keyCode":32,"modifierFlags":786432},"topRight":{"keyCode":34,"modifierFlags":786432}},"version":"96"}
```

**Reply by rxhanson** [reply]
`Center` is a little different in that it's not considered a "half" action, so it's not included in the set of actions that will cycle sizes. Since Center doesn't change the size of the window, what is your expected behavior if this existed? Would the window grow or shrink to a rectangle that is 2/3 or 1/3 the area of the screen upon repeated actions? 

I don't mind merging a good pull request that adds this as a new Terminal command configuration.

Alternatively, I can show you how this would be created in [Rectangle Pro](https://rectangleapp.com/pro) (paid), in the Custom Size/Position builder there, if desired.

**Reply by danielhoherd** [reply]
Sorry, maybe I misunderstood the point of this discussion? My reading of it is that people were looking for a behavior for center-snapping that is just like other snaps when the keystroke is repeated. When I snap right and repeat the keystroke, the window cycles between my configured areas, like 1/2, 1/3, 1/4. I am looking for that same behavior for center. I want to be able to press cmd-opt-c repeatedly and get a center 1/2, 1/3, 1/4 snap.

**Reply by rxhanson** [reply]
No worries. This discussion is about "Center Half" rather than "Center". Center Half is 1/2 width, full height (or rotated for portrait oriented displays). Center will move a window but won't resize it, so there's no existing notion of 1/2 size to base the cycle on, unless we were to make an assumption that you wanted something like 1/2 the area of the screen, centered (which would be a different action than what exists).
